chain/tsresolver/tipset_resolver.go

```diff
@@ -0,0 +1,164 @@
+	f3     F3
+}
+
+var _ TipSetResolver = (*tipSetResolver)(nil)
+
+func NewTipSetResolver(loader TipSetLoader, f3 F3) TipSetResolver {
+	return &tipSetResolver{
+		loader: loader,
+		f3:     f3,
+	}
+}
+
+// ResolveEthBlockSelector resolves an Ethereum block selector string to a TipSet.
+//
+// The selector can be one of:
+//   - "pending": the chain head
+//   - "latest": the TipSet with the latest executed messages (head - 1)
+//   - "safe": the TipSet with messages executed at least eth.SafeEpochDelay (30) epochs ago or the
+//     latest F3 finalized TipSet
+//   - "finalized": the TipSet with messages executed at least policy.ChainFinality (900) epochs ago
+//     or the latest F3 finalized TipSet
+//   - a decimal block number: the TipSet at the given height
+//   - a 0x-prefixed hex block number: the TipSet at the given height
+//
+// If a specific block number is specified and `strict` is true, an error is returned if the block
+// number resolves to a null round, otherwise in the case of a null round the first non-null TipSet
+// immediately before the null round is returned.
+func (tsr *tipSetResolver) ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
+	switch selector {
+	case EthBlockSelectorEarliest:
+		return nil, fmt.Errorf(`block param "%s" is not supported`, EthBlockSelectorEarliest)
+
+	case EthBlockSelectorPending:
+		return tsr.loader.GetHeaviestTipSet(), nil
+
+	case EthBlockSelectorLatest:
+		// head - 1 because we're always one behind for Eth compatibility due to deferred execution
+		return tsr.loader.LoadTipSet(ctx, tsr.loader.GetHeaviestTipSet().Parents())
+
+	case EthBlockSelectorSafe, EthBlockSelectorFinalized:
+		defaultDelay := policy.ChainFinality
+		if selector == EthBlockSelectorSafe {
+			defaultDelay = ethtypes.SafeEpochDelay
+		}
+
+		head := tsr.loader.GetHeaviestTipSet()
+		latestHeight := head.Height() - 1 // always one behind for Eth compatibility due to deferred execution
+		defaultHeight := latestHeight - defaultDelay
+
+		if f3TipSet, err := tsr.getF3FinalizedTipSet(ctx); err != nil {
+			return nil, err
+		} else if f3TipSet != nil && f3TipSet.Height() > defaultHeight {
+			// return the parent of the finalized tipset (deferred execution, eth cares about t)
+			return tsr.loader.LoadTipSet(ctx, f3TipSet.Parents())
+		} // else F3 is disabled, not running, or behind the default safe or finalized height
+
+		ts, err := tsr.loader.GetTipsetByHeight(ctx, defaultHeight, head, true)
+		if err != nil {
+			return nil, xerrors.Errorf("loading tipset at height %v: %w", defaultHeight, err)
+		}
+		return ts, nil
+
+	default:
+		// likely an 0x hex block number or a decimal block number
+		return tsr.resolveEthBlockNumberSelector(ctx, selector, strict)
+	}
+}
+
+func (tsr *tipSetResolver) getF3FinalizedTipSet(ctx context.Context) (*types.TipSet, error) {
+	if !tsr.f3.IsEnabled() {
+		return nil, nil
+	}
+	if running, _ := tsr.f3.IsRunning(); !running {
```

Comment:
👍

I see this now on the godocs:

> IsRunning returns true if gpbft is running Used mainly for testing purposes

Does that mean it doesn't have much utility after launch? Should an API consumer just be relying on error types because "it's not running" is an abnormal state?

Reply:
I have a follow-up question to this, will comment inline again on the latest iteration